rfcs/text/0000-metadata-integration-test.md

```diff
@@ -0,0 +1,78 @@
+
+
+Design
+
+1. Bundles with new metadata schema changes are uploaded to the production environment with a new schema label
+1. Observing the new schema label, the DSS stores the new bundles and makes them available with the `GET /v1/bundles` endpoint with the bundle `uuid` or `(uuid, version)`, but does not index nor does it release subscription notifications for them.
+1. Via a schema integration test system, a sample of the uploaded bundles are copied from the production DSS to the integration DSS testing environment.
+1. The integration DSS deployment stores the bundles, indexes their metadata, and releases bundle events to downstream systems in the integration environment with subscriptions.
+1. The schema integration test checks the results and passes if systems downstream of the integration DSS have correctly processed the new bundles; it fails otherwise.*
+1. If the schema integration test passes, the new bundles in the production DSS can be released downstream by issuing a new bundle version of each without the new schema label. This process could be automated.
+1. If the schema test does not pass, the data is not released and development teams are notified to resolve the issue and rerun the test.
+
+![Integration test design](../images/0000-integration-test.png)
+
+*Integration test data flow: test path is shown in red, production release in blue, integration tooling in dashed lines; edges are labeled with corresponding step numbers in the design above*
+
+\* To estimate the impact of schema changes on third party applications, the schema integration test system could run critical [data consumer vignettes](https://github.com/HumanCellAtlas/data-consumer-vignettes).
+
+### Pros:
+
+* Bundles only have to be uploaded once
+* Bundles can be uploaded directly to the production environment
+* Tests exactly what bundles downstream systems will see; does not require complicated fake metadata generation code in tests
+* Data in storage is immediately available after upload via the `GET /v1/bundles` endpoint
+* Downstream systems are protected from new schema changes until it can be confirmed that they are able to handle the changes
+* Does not require replaying subscription notifications
+
+### Cons:
+
+* Does not test new bundles against the ingestion pipeline, this is assumed to "just work"
+* Does not cover third party systems, which may be consuming from the DSS
+* New bundles are not available in the data browser until tests for the entire system have passed
+* There is no hard guarantee that downstream production systems will be able to handle new bundles: there could be critical updates in the integration environment which have not yet reached production
+* In the case that the integration test does not pass, developers and wranglers will have to exercise care in propagating fixes to production before releasing the new bundles downstream of the production DSS
+
+## Discussion:
```

Comment:
Glad you brought this point up. I agree and think validating the client's "contract" on the schema is a subsequent logical step after we have integration testing.

In the future, it would be great to provide users with something like a schema contract library to do exactly what you say (see [slides](https://docs.google.com/presentation/d/1Y28vT6PbPDC6ublsZIIMNTvianCxVLlt4M5sypwX2wE/edit#slide=id.g4339acaf8a_0_66) and [sample implementation](https://github.com/HumanCellAtlas/metadata-migration#-contracts)).

Importantly, we need proper integration testing too to understand which clients will be broken by a given schema change before we send it out to clients and to check that our stated dependencies on the schema are correct.

Reply:
+1 to all this